src/batchmap.rs

```diff
@@ -0,0 +1,627 @@
+    pub watermark: DateTime<Utc>,
+    /// Time of the element as seen at source or aligned after a reduce operation.
+    pub event_time: DateTime<Utc>,
+    /// ID is the unique id of the message to be sent to the Batch Map.
+    pub id: String,
+    /// Headers for the message.
+    pub headers: HashMap<String, String>,
+}
+
+impl From<proto::BatchMapRequest> for Datum {
+    fn from(sr: proto::BatchMapRequest) -> Self {
+        Self {
+            keys: sr.keys,
+            value: sr.value,
+            watermark: shared::utc_from_timestamp(sr.watermark),
+            event_time: shared::utc_from_timestamp(sr.event_time),
+            id: sr.id,
+            headers: sr.headers,
+        }
+    }
+}
+/// Message is the response struct from the [`Mapper::map`] .
+#[derive(Debug, PartialEq)]
+pub struct Message {
+    /// Keys are a collection of strings which will be passed on to the next vertex as is. It can
+    /// be an empty collection.
+    pub keys: Option<Vec<String>>,
+    /// Value is the value passed to the next vertex.
+    pub value: Vec<u8>,
+    /// Tags are used for [conditional forwarding](https://numaflow.numaproj.io/user-guide/reference/conditional-forwarding/).
+    pub tags: Option<Vec<String>>,
+}
+
+/// Represents a message that can be modified and forwarded.
+impl crate::batchmap::Message {
+    /// Creates a new message with the specified value.
+    ///
+    /// This constructor initializes the message with no keys, tags, or specific event time.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - A vector of bytes representing the message's payload.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use numaflow::batchmap::Message;
+    /// let message = Message::new(vec![1, 2, 3, 4]);
+    /// ```
+    pub fn new(value: Vec<u8>) -> Self {
+        Self {
+            value,
+            keys: None,
+            tags: None,
+        }
+    }
+    /// Marks the message to be dropped by creating a new `Message` with an empty value and a special "DROP" tag.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use numaflow::batchmap::Message;
+    /// let dropped_message = Message::message_to_drop();
+    /// ```
+    pub fn message_to_drop() -> crate::batchmap::Message {
+        crate::batchmap::Message {
+            keys: None,
+            value: vec![],
+            tags: Some(vec![crate::batchmap::DROP.to_string()]),
+        }
+    }
+
+    /// Sets or replaces the keys associated with this message.
+    ///
+    /// # Arguments
+    ///
+    /// * `keys` - A vector of strings representing the keys.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///  use numaflow::batchmap::Message;
+    /// let message = Message::new(vec![1, 2, 3]).keys(vec!["key1".to_string(), "key2".to_string()]);
+    /// ```
+    pub fn keys(mut self, keys: Vec<String>) -> Self {
```

Comment:
The conventions for builder pattern is to have a `with_` prefix for methods eg. `pub fn with_keys(mut self, keys: Vec<String>) -> Self`

Reply:
I went ahead with the style followed in the current map code
https://github.com/numaproj/numaflow-rs/blob/db74f62cbb8eb3df44533cf4b63fbe81f4379750/src/map.rs#L99
Based on what we want to go forward with it would be good to keep it consistent throughout then.